rfcs/records.md

```diff
@@ -0,0 +1,201 @@
+# Records
+
+## Summary
+
+Introduces a new data type, record, which is essentially a dictionary with fixed structure.
+
+## Motivation
+
+Lua has tables as the only data structure. Tables are very versatile - they can model arrays, dictionaries, tuples, objects, etc.
+With metatables, they can also model rich OOP patterns like inheritance or properties.
+
+This flexibility comes at a cost - using tables for everything means that for every specific usecase tables are likely suboptimal.
+
+A lot of data in Luau programs uses tables as objects - tables with more or less rigid set of string keys. Sometimes these objects have
+metatables that define methods, sometimes they overload operators, sometimes they are just used to store data.
+
+Tables aren't perfect for objects; this proposal is motivated by the following deficiencies in using tables-as-objects (in rough priority order):
+
+1. Memory overhead. An object with 6 fields takes `8+56+8+8*32=328` bytes of storage.
+This is because tables have a large header, the hash portion is rounded to a power of two, and every entry has key and value despite the fixed object structure.
+2. Type system compatibility. While Luau type checker can type tables, including ones with rigid structure, it falls short in OOP scenarios because it's very difficult
+to associate methods with table structure in idiomatic OOP in Luau with tables. Today the situation is especially dire because each method gets its own inferred self type
+(something that is likely to change), and it's impossible to specify a table-with-metatable type via obvious type syntax.
```

Comment:
I'm not sure actually that partly built object is a problem with records. It is with tables because your shape at the type level (and runtime) mutates through assignments. Records don't do that. Constructing the record without specifying all fields is going to produce a type error if fields aren't optional.

Reply:
Hmm, the nasty cases are things like
```
local MT = {}
function MT:f() doSomethingWith(self.x) end
function MT.new()
  local result = setmetatable(MT, {})
  result.x = 5
  self:f()
  result.y = 7
end
```
and if we're prepared to infer that `x` and `y` are of type `number?` then between tables with shared self, and records there's not much daylight.

Reply:
Yeah that's exactly the nasty case with tables, but records don't have this problem because you construct a fully built record object before using it.

Reply:
I've added a paragraph around this.